pulsar/transaction_impl.go

```diff
@@ -0,0 +1,232 @@
+	}()
+	transaction.log = tcClient.log.SubLogger(log.Fields{})
+	return transaction
+}
+
+func (txn *transaction) GetState() TxnState {
+	return txn.state
+}
+
+func (txn *transaction) Commit(ctx context.Context) error {
+	if !(atomic.CompareAndSwapInt32((*int32)(&txn.state), int32(Open), int32(Committing)) || txn.state == Committing) {
+		return newError(InvalidStatus, "Expect transaction state is Open but "+txn.state.string())
+	}
+
+	//Wait for all operations to complete
+	select {
+	case <-txn.opsFlow:
+	case <-time.After(txn.opTimeout):
+		return newError(TimeoutError, "There are some operations that are not completed after the timeout.")
+	}
+	//Send commit transaction command to transaction coordinator
+	err := txn.tcClient.endTxn(&txn.txnID, pb.TxnAction_COMMIT)
+	if err == nil {
+		atomic.StoreInt32((*int32)(&txn.state), int32(Committed))
+	} else {
+		if err.(*Error).Result() == Result(pb.ServerError_TransactionNotFound) || err.(*Error).Result() == InvalidStatus {
+			atomic.StoreInt32((*int32)(&txn.state), int32(Errored))
+			return err
+		}
+		txn.opsFlow <- true
+	}
+	return err
+}
+
+func (txn *transaction) Abort(ctx context.Context) error {
+	if !(atomic.CompareAndSwapInt32((*int32)(&txn.state), int32(Open), int32(Aborting)) || txn.state == Aborting) {
+		return newError(InvalidStatus, "Expect transaction state is Open but "+txn.state.string())
+	}
+
+	//Wait for all operations to complete
+	select {
+	case <-txn.opsFlow:
+	case <-time.After(txn.opTimeout):
+		return newError(TimeoutError, "There are some operations that are not completed after the timeout.")
+	}
+	//Send abort transaction command to transaction coordinator
+	err := txn.tcClient.endTxn(&txn.txnID, pb.TxnAction_ABORT)
+	if err == nil {
+		atomic.StoreInt32((*int32)(&txn.state), int32(Aborted))
+	} else {
+		if err.(*Error).Result() == Result(pb.ServerError_TransactionNotFound) ||
+			err.(*Error).Result() == Result(pb.ServerError_InvalidTxnStatus) {
+			atomic.StoreInt32((*int32)(&txn.state), int32(Errored))
+		} else {
+			txn.opsFlow <- true
+		}
+	}
+	return err
+}
+
+func (txn *transaction) registerSendOrAckOp() error {
+	if atomic.AddInt32(&txn.opsCount, 1) == 1 {
+		//There are new operations that not completed
+		select {
+		case <-txn.opsFlow:
+			return nil
+		case <-time.After(txn.opTimeout):
+			if _, err := txn.checkIfOpen(); err != nil {
+				return err
+			}
+			return newError(TimeoutError, "Failed to get the semaphore to register the send/ack operation")
+		}
+	}
+	return nil
+}
+
+func (txn *transaction) endSendOrAckOp(err error) {
+	if err != nil {
+		atomic.StoreInt32((*int32)(&txn.state), int32(Errored))
+	}
+	if atomic.AddInt32(&txn.opsCount, -1) == 0 {
+		//This means there are not pending send/ack requests
+		txn.opsFlow <- true
+	}
+}
+
+func (txn *transaction) registerProducerTopic(topic string) error {
+	isOpen, err := txn.checkIfOpen()
+	if !isOpen {
+		return err
+	}
+	_, ok := txn.registerPartitions[topic]
```

Comment:
This map also needs to be protected with a lock, right?

Reply:
Yeah, we should check the map in the lock again.